panic: flow: no bytes in account to release, v1.1 alpha

**BUG REPORT**

1. Please supply the header (i.e. the first few lines) of your most recent
   log file **for each node in your cluster**. On most unix-based systems
   running with defaults, this boils down to the output of

     grep -F '[config]' cockroach-data/logs/cockroach.log

   When log files are not available, supply the output of `cockroach version`
   and all flags/environment variables passed to `cockroach start` instead.

```
I170828 16:02:37.934602 1 util/log/clog.go:975  [config] file created at: 2017/08/28 16:02:37
I170828 16:02:37.934602 1 util/log/clog.go:975  [config] running on machine: ashley
I170828 16:02:37.934602 1 util/log/clog.go:975  [config] binary: CockroachDB CCL v1.1-alpha.20170817 (linux amd64, built 2017/08/17 22:29:49, go1.8.3)
I170828 16:02:37.934602 1 util/log/clog.go:975  [config] arguments: [cockroach start --insecure --host=localhost --http-port 8181]
```

Full logs: https://gist.github.com/d4l3k/bc5c15060e0a7916232cdc41b4fa78ff
Running with only a single node on localhost.
Binary installed via https://aur.archlinux.org/packages/cockroachdb-bin/

Occured with version `1.1_alpha.20170803` and `1.1_alpha.20170817`.

2. Please describe the issue you observed:

- What did you do?

I launched a server and then attempted to query the DB instance via gorm/sql.
```
$ cockroach sql -d pok --insecure
# Welcome to the cockroach SQL interface.
# All statements must be terminated by a semicolon.
# To exit: CTRL + D.
#
# Server version: CockroachDB CCL v1.1-alpha.20170817 (linux amd64, built 2017/08/17 22:29:49, go1.8.3) (same version as client)
# Cluster ID: 862f1d8f-1c5d-4553-bf10-ed1bba297c26
#
# Enter \? for a brief introduction.
#
root@:26257/pok> SELECT * FROM models WHERE (models.deleted_at IS NULL) AND ((models.domain_id = 'luk.ai')) ORDER BY updated_at DESC LIMIT 15 OFFSET 0;
driver: bad connection
connection lost; opening new connection: all session settings will be lost
error retrieving the transaction status: read tcp 127.0.0.1:34568->127.0.0.1:26257: read: connection reset by peer
connection lost; opening new connection: all session settings will be lost
error retrieving the database name: dial tcp [::1]:26257: getsockopt: connection refused
```

- What did you expect to see?

It return the results.

- What did you see instead?

It panicked with

```
panic: flow: no bytes in account to release, current 1456, free 2064

goroutine 409 [running]:
github.com/cockroachdb/cockroach/pkg/sql/mon.(*BytesMonitor).ShrinkAccount(0xc421134d80, 0x7ff6ca23c918, 0xc4200144c8, 0xc42116fb68, 0x810)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/mon/bytes_usage.go:471 +0x1d5
github.com/cockroachdb/cockroach/pkg/sql/mon.(*BoundAccount).Shrink(0xc42116fb68, 0x7ff6ca23c918, 0xc4200144c8, 0x810)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/mon/bytes_usage.go:546 +0x52
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowContainer).PopFirst(0xc42116fb00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/row_container.go:318 +0x1f2
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sortTopKStrategy).Execute(0xc4202c4c90, 0x7ff6ca1f2930, 0xc4211a0c60, 0xc421164900, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorterstrategy.go:235 +0x284
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*sorter).Run(0xc421164900, 0x7ff6ca1f2930, 0xc4211a0c60, 0xc42116f888)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sorter.go:150 +0x652
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:360 +0x332
```

It's very reproducible. I just have to launch the server and run that SQL command.

--- a/pkg/sql/distsqlrun/row_container.go
+++ b/pkg/sql/distsqlrun/row_container.go
@@ -166,7 +166,7 @@ func (mc *memRowContainer) Pop() interface{} { panic("unimplemented") }
 
 // MaybeReplaceMax replaces the maximum element with the given row, if it is smaller.
 // Assumes InitMaxHeap was called.
-func (mc *memRowContainer) MaybeReplaceMax(row sqlbase.EncDatumRow) error {
+func (mc *memRowContainer) MaybeReplaceMax(ctx context.Context, row sqlbase.EncDatumRow) error {
 	max := mc.At(0)
 	cmp, err := row.CompareToDatums(&mc.datumAlloc, mc.ordering, mc.evalCtx, max)
 	if err != nil {
@@ -178,7 +178,10 @@ func (mc *memRowContainer) MaybeReplaceMax(row sqlbase.EncDatumRow) error {
 			if err := row[i].EnsureDecoded(&mc.datumAlloc); err != nil {
 				return err
 			}
-			max[i] = row[i].Datum
+			mc.scratchRow[i] = row[i].Datum
+		}
+		if err := mc.Replace(ctx, 0, mc.scratchRow); err != nil {
+			return err
 		}
 		heap.Fix(mc, 0)
 	}

--- a/pkg/sql/distsqlrun/row_container_test.go
+++ b/pkg/sql/distsqlrun/row_container_test.go
@@ -75,7 +75,7 @@ func TestRowContainerReplaceMax(t *testing.T) {
 	mc.InitMaxHeap()
 	// Replace some of the rows with large rows.
 	for i := 0; i < 1000; i++ {
-		err := mc.MaybeReplaceMax(makeRow(rng.Intn(10000), rng.Intn(100)))
+		err := mc.MaybeReplaceMax(ctx, makeRow(rng.Intn(10000), rng.Intn(100)))
 		if err != nil {
 			t.Fatal(err)
 		}

--- a/pkg/sql/distsqlrun/row_container_test.go
+++ b/pkg/sql/distsqlrun/row_container_test.go
@@ -1,0 +1,88 @@
+// Copyright 2017 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package distsqlrun
+
+import (
+	"math"
+	"testing"
+
+	"golang.org/x/net/context"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/mon"
+	"github.com/cockroachdb/cockroach/pkg/sql/parser"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+	"github.com/cockroachdb/cockroach/pkg/util/encoding"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+)
+
+// TestRowContainerReplaceMax verifies that MaybeReplaceMax correctly adjusts
+// the memory accounting.
+func TestRowContainerReplaceMax(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	rng, _ := randutil.NewPseudoRand()
+
+	evalCtx := parser.NewTestingEvalContext()
+	defer evalCtx.Stop(ctx)
+
+	typeInt := sqlbase.ColumnType{SemanticType: sqlbase.ColumnType_INT}
+	typeStr := sqlbase.ColumnType{SemanticType: sqlbase.ColumnType_STRING}
+
+	makeRow := func(intVal int, strLen int) sqlbase.EncDatumRow {
+		var b []byte
+		for i := 0; i < strLen; i++ {
+			b = append(b, 'a')
+		}
+		return sqlbase.EncDatumRow{
+			sqlbase.DatumToEncDatum(typeInt, parser.NewDInt(parser.DInt(intVal))),
+			sqlbase.DatumToEncDatum(typeStr, parser.NewDString(string(b))),
+		}
+	}
+
+	m := mon.MakeUnlimitedMonitor(
+		context.Background(), "test", mon.MemoryResource, nil, nil, math.MaxInt64,
+	)
+	defer m.Stop(ctx)
+
+	var mc memRowContainer
+	mc.initWithMon(
+		sqlbase.ColumnOrdering{{ColIdx: 0, Direction: encoding.Ascending}},
+		[]sqlbase.ColumnType{typeInt, typeStr}, evalCtx, &m,
+	)
+	defer mc.Close(ctx)
+
+	// Initialize the heap with small rows.
+	for i := 0; i < 1000; i++ {
+		err := mc.AddRow(ctx, makeRow(rng.Intn(10000), rng.Intn(10)))
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	mc.InitMaxHeap()
+	// Replace some of the rows with large rows.
+	for i := 0; i < 1000; i++ {
+		err := mc.MaybeReplaceMax(makeRow(rng.Intn(10000), rng.Intn(100)))
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	// Now pop the rows, which shrinks the memory account according to the current
+	// row sizes. If we did not account for the larger rows, this will panic.
+	for mc.Len() > 0 {
+		mc.PopFirst()
+	}
+}

--- a/pkg/sql/distsqlrun/sorterstrategy.go
+++ b/pkg/sql/distsqlrun/sorterstrategy.go
@@ -215,7 +215,7 @@ func (ss *sortTopKStrategy) Execute(ctx context.Context, s *sorter) error {
 			}
 			// Replace the max value if the new row is smaller, maintaining the
 			// max-heap.
-			if err := ss.rows.MaybeReplaceMax(row); err != nil {
+			if err := ss.rows.MaybeReplaceMax(ctx, row); err != nil {
 				return err
 			}
 		}

--- a/pkg/sql/sqlbase/row_container_test.go
+++ b/pkg/sql/sqlbase/row_container_test.go
@@ -69,6 +69,8 @@ func TestRowContainer(t *testing.T) {
 						}
 					}
 				}
+				rc.Close(context.Background())
+				m.Stop(context.Background())
 			}
 		}
 	}

